Allow reading nonce if it was included in header

jwhitlock opened the pull request:
If the nonce was generated before the CSP headers were set, then allow reading it with ``request.csp_nonce``.

If the CSP headers were set with no nonce, then continue raising ``CSPNonceError`` when reading it as a string. If read as a boolean (`if request.csp_nonce`), then return `False`.

This will allow other middleware like [django-debug-toolbar](https://github.com/django-commons/django-debug-toolbar) to alter the response after the CSP middleware runs, fixing #268.

robhudson commented on csp/middleware.py:
```diff
@@ -29,6 +29,11 @@ class PolicyParts:
 
+class FalseLazyObject(SimpleLazyObject):
+    def __bool__(self) -> bool:
+        return False
```
Do you think a test just for this would be worthwhile?

jwhitlock replied:
yes, see below

robhudson replied:
I saw those. I meant more on its own... not wrapped up in the middleware. More a question if you think that covers it well enough or should there be separate unit tests to test these isolated. I'm fine with what's here in this PR.

jwhitlock replied:
Ah. No, I think this is testing at the right level. I don't expect others to use this class. I could rename to `_FalseLazyObject` to make it clearer that it is private.

robhudson commented on csp/middleware.py:
```diff
@@ -95,7 +100,8 @@ def process_response(self, request: HttpRequest, response: HttpResponseBase) ->
         # this has security implications.
-        setattr(request, "csp_nonce", SimpleLazyObject(self._csp_nonce_post_response))
+        if getattr(request, "_csp_nonce", None) is None:
+            setattr(request, "csp_nonce", FalseLazyObject(self._csp_nonce_post_response))
```
This is a significant change...

- If the nonce was used during the request it should then be set and not `None`. In which case, we could continue to read it after the middleware just fine.
- If the none was NOT used during the request, we update it with this so a boolean check returns False and any attempt to access it raises the `CSPNonceError`.

That seems like a nice approach. We should document this change in behavior.

jwhitlock replied:
I agree, I'll update `docs/nonce.rst`. I think that would be a good place for my advice for middleware that needs the nonce.

jwhitlock commented on csp/tests/test_middleware.py:
```diff
@@ -160,10 +160,21 @@ def test_nonce_regenerated_on_new_request() -> None:
     request = rf.get("/")
     mw.process_request(request)
     mw.process_response(request, HttpResponse())
+    assert bool(getattr(request, "csp_nonce", True)) is False
```
@robhudson this is the test for `bool(request.csp_nonce) is False`

jwhitlock commented on csp/tests/test_middleware.py:
```diff
@@ -160,10 +160,21 @@ def test_nonce_regenerated_on_new_request() -> None:
+    mw.process_request(request)
+    nonce = str(getattr(request, "csp_nonce"))
+    mw.process_response(request, HttpResponse())
+    assert bool(getattr(request, "csp_nonce", False)) is True
```
and this is the test for `bool(request.csp_nonce) is True`

jwhitlock reviewed (commented):
Thanks @robhudson, I'll add some docs. And maybe another middleware that always generates the nonce...

jwhitlock commented:
I added a commit:

* I added ``CSPMiddlewareAlwaysGenerateNonce``, which is like ``CSPMiddleware``, except it always generates a nonce. This could be used when using `django-debug-toolbar` or similar middlewares.
* I updated ``nonce.rst`` with the new behavior of `request.csp_nonce`, and when to use ``CSPMiddlewareAlwaysGenerateNonce`` or other method.
* I updated ``installation.rst`` to refer to ``nonce.rst`` for more complex init.

I meant to do this over two commits, `git` fail 🤦

robhudson commented:
The `CSPMiddlewareAlwaysGenerateNonce` gave me pause. But I think some would rather use that in certain situations. The nonce would be added to the header even if not used in the content. I can't see any security implications for this. I actually think this is the generally expected functionality. I think it *not* being in the header if not used causes some confusion, although I believe it's a nice feature.

I like these changes.
